### PR TITLE
feat(bindings): useStyles hook

### PR DESCRIPTION
**Pull request by layershifter**
This PR adds `useStyles()` to reuse theming capabilities in custom and Fluent components and have the same API for public and private usage.

### Usage

The example below assumes a component called `<Text>` will be used this way:

```tsx
type TextComponentProps = {
  className?: string
  color?: string
}

const Text: React.FunctionComponent<TextComponentProps> = props => {
  const { className, children, color } = props

  const [classes] = useStyles('Text', {
    className: 'ui-text',
    mapPropsToStyles: () => ({ color }),
  })

  return <span className={classes.root}>{children}</span>
}
```

### Reference

```tsx
const [classes] = useStyles(
  displayName: string,
  options: UseStylesOptions<Props>,
)
```

- `displayName` - a component name to lookup in theme
- `options.className` - optional, a special class name that will be always added to the `root` slot
- `options.mapPropsToStyles` - optional, a set of props that will be passed style functions, only primitives are allowed
- `options.rtl` - optional, sets RTL mode

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -9,11 +9,13 @@ const packagePaths = lernaAliases({ sourceDirectory: false })
...
-    !excludedPackages.includes(packageName) &&
-    !fs.existsSync(path.join(packagePaths[packageName], 'just.config.ts')),
-)
+const projects = Object.keys(packagePaths)
+  .filter(
+    packageName =>
+      !excludedPackages.includes(packageName) &&
+      !fs.existsSync(path.join(packagePaths[packageName], 'just.config.ts')),
+  )
+  .map(packageName => packagePaths[packageName])
```
Comment (layershifter) [inline jest.config.js]: Move me out a separate PR

File: package.json [hunk 2]
```diff
@@ -1,5 +1,8 @@
 {
   "private": true,
+  "engines": {
+    "node": ">= 10.0.0"
+  },
```
Comment (layershifter) [inline package.json]: Move me out a separate PR

File: packages/react-bindings/src/hooks/useStyles.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  theme: emptyTheme,
+  _internal_resolvedComponentVariables: {},
+}
+
+const useStyles = <StyleProps extends PrimitiveProps>(
+  displayName: string,
+  options: UseStylesOptions<StyleProps>,
+): [GetStylesResult['classes'], GetStylesResult['styles']] => {
+  const context: StylesContextValue<{ renderRule: RendererRenderRule }> =
+    React.useContext(ThemeContext) || defaultContext
```
Comment (jdhuntington) [inline packages/react-bindings/src/hooks/useStyles.ts:53]: Any ideas on how we could decouple this call from the concrete theme? It would be nice to be able to reuse the base components with a different design system (which would likely have a separate theme shape).

Comment (layershifter) [reply packages/react-bindings/src/hooks/useStyles.ts:53]: This is next stage goal 🙇‍♂️

This is actually even not `ThemeContext` as it contains everything, like `renderer` and `disableAnimations`. Currently we can't we have multiple contexts as the most of our components are class components 🙃

I have a naive idea that we will be able to move out `renderer` to a separate context and use the same interface different engines, https://codesandbox.io/embed/t-rex-proto-1w8bp

File: packages/react/src/components/Box/Box.tsx [hunk 4]
```diff
@@ -8,40 +10,70 @@ import {
...
+      styles,
+      variables,
+    }),
+  })
 
-  render(config, props) {
-    const { ElementType, classes, unhandledProps } = config
-    const { children, content } = props
+  const ElementType = getElementType(props)
+  const unhandledProps = getUnhandledProps(Object.keys(Box.propTypes) as any, props)
```
Comment (layershifter) [inline packages/react/src/components/Box/Box.tsx]: These changes should be reverted

File: packages/react/src/components/Debug/FiberNavigator.ts [hunk 5]
```diff
@@ -324,13 +324,33 @@ class FiberNavigator {
...
+      thing.memoizedState.current.fluentUIDebug
+    ) {
+      return thing.memoizedState.current
+    }
+
+    if (thing === null || thing.next === null) {
+      return null
+    }
+
+    return this.findMemoizedState(thing.next)
```
Comment (layershifter) [inline packages/react/src/components/Debug/FiberNavigator.ts]: I suggest to rename to `findDebugHookState`

File: packages/react/src/components/Debug/FiberNavigator.ts [hunk 6]
```diff
@@ -358,10 +378,6 @@ class FiberNavigator {
     return !!fiberNav && fiberNav.instance === this.instance
   }
 
-  usesHook(name) {
```
Comment (layershifter) [inline packages/react/src/components/Debug/FiberNavigator.ts]: Cruft

File: packages/styles/src/types.ts [hunk 7]
```diff
@@ -201,7 +201,7 @@ export interface ComponentStyleFunctionParam<
 }
 
 export type ComponentSlotStyleFunction<TProps = {}, TVars = {}> = (
-  styleParam?: ComponentStyleFunctionParam<TProps, TVars>,
+  styleParam: ComponentStyleFunctionParam<TProps, TVars>,
```
Comment (layershifter) [inline packages/styles/src/types.ts]: Let's ship this and changes in tests as a separate PR

File: packages/react/src/components/Debug/FiberNavigator.ts [hunk 8]
```diff
@@ -324,13 +324,33 @@ class FiberNavigator {
...
+
+    if (this.isFunctionComponent) {
+      // assumes functional component w/useRef
+      return this.findMemoizedState(this.__fiber.memoizedState)
+    }
+
+    return null
+  }
+
+  findMemoizedState(thing) {
```
Comment (layershifter) [inline packages/react/src/components/Debug/FiberNavigator.ts]: `node` may be?

```
Hooks state is represented by a recursive structure where:
- `memoizedState` is a current value if applicable
- `next` is next hook in order
```

Let's also add a descriptive comment

Comment (mnajdova) [reply packages/react/src/components/Debug/FiberNavigator.ts]: thing is awesome :D

**Review by mnajdova — APPROVED** [review]
